File: 02-linked-lists/03_detect_cycle/detect_cycle.py
"""
Problem: Linked List Cycle Detection
LeetCode: #141 (Easy)

Approach: Floyd's Tortoise and Hare
Time Complexity: O(N)
Space Complexity: O(1)
"""

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def has_cycle(head: Optional[ListNode]) -> bool:
    if not head or not head.next:
        return False
    slow, fast = head, head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
    return False

if __name__ == "__main__":
    print("[Python] Detect Cycle Test")
    n1 = ListNode(3)
    n2 = ListNode(2)
    n3 = ListNode(0)
    n4 = ListNode(-4)
    n1.next = n2
    n2.next = n3
    n3.next = n4
    n4.next = n2 # Cycle

    assert has_cycle(n1) is True
    print("Cycle detected correctly!")

File: 02-linked-lists/03_detect_cycle/DetectCycle.java
/**
 * Problem: Linked List Cycle Detection
 * LeetCode: #141 (Easy)
 * 
 * Approach: Floyd's Tortoise and Hare
 * Time Complexity: O(N)
 * Space Complexity: O(1)
 */

public class DetectCycle {
    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
    }

    public static boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) return false;

        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) return true;
        }
        return false;
    }

    public static void main(String[] args) {
        System.out.println("[Java] Detect Cycle Test");
        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(3);
        head.next.next.next = head; // Cycle to head

        System.out.println("Has Cycle: " + hasCycle(head));
    }
}
